iceoryx2-cal/src/zero_copy_connection/used_chunk_list.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use std::{
+    alloc::Layout,
+    sync::atomic::{AtomicBool, Ordering},
+};
+
+use iceoryx2_bb_elementary::{
+    math::align_to,
+    owning_pointer::OwningPointer,
+    relocatable_container::RelocatableContainer,
+    relocatable_ptr::{PointerTrait, RelocatablePointer},
+};
+use iceoryx2_bb_log::{fail, fatal_panic};
+
+use crate::shm_allocator::PointerOffset;
+
+pub type UsedChunkList = details::UsedChunkList<OwningPointer<AtomicBool>>;
+pub type RelocatableUsedChunkList = details::UsedChunkList<RelocatablePointer<AtomicBool>>;
+
+pub mod details {
+    use std::fmt::Debug;
+
+    use iceoryx2_bb_elementary::owning_pointer::OwningPointer;
+
+    use super::*;
+
+    #[derive(Debug)]
+    #[repr(C)]
+    pub struct UsedChunkList<PointerType: PointerTrait<AtomicBool>> {
+        data_ptr: PointerType,
+        capacity: usize,
```

Comment:
Food for thought. I guess `capacity` will never change after initialization. A type only with a `new` function and `Deref` would bake this invariant into the type system.

Reply:
Yes you are right but I will not fix this in this PR I created an issue and would tackle this in a follow up PR. (#142 )